### DISCUSSION TITLE
Can this work for Godot 4+ games with multiplayer and open world chunks?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [javier-adrian/voxel-gds#2] [Q&A]
Test chunks link: https://github.com/WithinAmnesia/ARPG/discussions/15

I'm trying to find a way to seamless load and unload chunks for a 2D multiplayer game project to make an open world with a working server using Godot 4.2.1.NET.

How can this work for multiplayer and what is needed for this to potentially work? What options can be used for chunk loading and unloading seamlessly in Godot 4.2.1.NET? Please give feedback.

**Comment by WithinAmnesia** [comment]
Here is an update since from before:
https://jonathaneeckhout.itch.io/jdungeon For the goal right now it is easy to play this but imagine it has seamless chunk loading and unloading. As instead of the black world boundary a new chunk loads in seamlessly. What are your suggestions and thoughts?

https://github.com/WithinAmnesia/ARPG/tree/ARPG-Infinite-Worlds Here is the 128x128 with 32x32 pixel tile chunk to test. How to start: Run 3-4 instances in the debug -> run multiple instances. Then run the first instance as Gateway. The second instance should run as Server. The third instance as Client. Make an account then log in. The fourth and more instances can be for client multi-boxing testing and work the same as the third instance.

My initial testing seems to feel even faster combat with the 128x128 chunk. This 128x128 chunk has the same amount of entities as the 256x256 chunk. For I moved all of the entities over into the 128x128 chunk. This 128x128 chunk is a good test for using the Infinite-Worlds using the BinarySerializer for Godot 4.2+ from Theraot: https://gist.github.com/theraot/31515e28e2d8bfea33f6c6d5bcd852f6 . There needs to be some testing how to make the chunks seamlessly load and unload.

**Reply by javier-adrian** [reply]
Apologies for the late response.

Most of the information given in this question is going way over my head, so it somewhat feels like I'm not the best person to ask this question to, I don't have a lot of experience in game development, let alone Godot. However, I'll still try my best.

> What are your suggestions and thoughts?

This project does chunk loading for a 3D game, so the code will have to be a whole lot different, but the concept is similar: Divide the world into chunks (it seems like you've already done this), store their positions, calculate the distance between the player and each chunk, then if the distance is low enough, load them.

Of course, for them to be seamless with each other, the _terrain_ of the chunks would have to be generated procedurally (get a FastNoiseLite, for each pixel tile: generate a terrain level by passing the position of the pixel tile to the FastNoiseLite. If the terrain level is low enough, the pixel tile is a river. If it is high enough it would be a cliffside or a mountain of some sorts.)

**Reply by javier-adrian** [reply]
If you want some help with your project, I'd be happy to help. Although college is eating my time away :(

**Reply by WithinAmnesia** [reply]
Awwwe that very nice of you.

**Reply by WithinAmnesia** [reply]
We just need some help with testing. It's okay I'm not good either but like we solve the puzzle everyone wins.

**Reply by WithinAmnesia** [reply]
Hey can you read this code better than me? Its from a person called Theraot they think it might help? "Theraot: https://gist.github.com/theraot/31515e28e2d8bfea33f6c6d5bcd852f6 . There needs to be some testing how to make the chunks seamlessly load and unload." they have more details here: https://gamedev.stackexchange.com/questions/209002/multiplayer-seamless-open-world-chunks

**Reply by javier-adrian** [reply]
[theraot/binary_serializer.gd](https://gist.github.com/theraot/31515e28e2d8bfea33f6c6d5bcd852f6) seems like it just saves data of a given type into a [StreamPeerBuffer](https://docs.godotengine.org/en/stable/classes/class_streampeerbuffer.html), which is from what I understand, a data buffer for network sessions. So the server would save data (things like the terrain data) using the `BinarySerializer` class into the `StreamPeerBuffer`, and then the clients would load those data from the `StreamPeerBuffer` also with the `BinarySerializer`. Basically, from what I understand, the `BinarySerializer` is just a tool to save and load data to the network session so that the server does all the generating of chunk data and the client only loads data from the server.

**Reply by WithinAmnesia** [reply]
So it works like make big things small into binary 0s and 1s and then good for transfer online server data and bandwidth time management (fast connection speeds / low game latency). As it works like the server reading reads the data and runs it through the server systems and with this code ~tool the server can make it small with binary 0s and 1s and the send it to the client. Then the client makes it from small binary 0s and 1s to big data it can read in the game and visa versa back and forth. So as to work for the player systems and read and process the the data fast between the server and the client so it helps get more data across when with small network connections? Its like compressor / de-compressor for helping the server and client / network communicate bigger data over a smaller network / bandwidth allowance type of thing? Its a way to optimize say server chunk loading and unloading?

**Reply by javier-adrian** [reply]
Yes, I literally missed the most important part where it translates it into binary data for ease of transfer. But that explanation seems to be right. So essentially you could optimize the transfer of chunk data from the server to the client.

**Reply by WithinAmnesia** [reply]
Oh okay thank you like I was guessing at first but the more you talked about it the more it made sense lol. Thank you for clarifying that. I could read the code but like I'm 1/3 the meaning it feels a bit fuzzy? I wonder how to plug that into some chunk loading demos now thank you. I'll keep trying to bring the community together around these:
~4 components/parts for smaller related questions.

   -  Chunk loading/World Partitioning
   - Multiplayer/Server
   - Navigation/Astar/Path-finding
   - Server/Chunk/Navigation systems working together/integration
   
 I wonder what do you like and are interested in hey? What are you interested in? As like maybe a little is better to work on with little time? What stuff do you like?

**Comment by WithinAmnesia** [comment]
Update: https://github.com/Zylann/voxelgame/issues/100#issuecomment-1955780686 We figured it out! Now its phase 3 with trying to convert the working multiplayer dynamic seamless chunk system to top down 2D. We are onto the next puzzle to solve now: https://github.com/Zylann/voxelgame/issues/101
